server/src/main/java/io/druid/client/CachingClusteredClient.java

```diff
@@ -219,14 +223,37 @@ public CachingClusteredClient(
     final List<TimelineObjectHolder<String, ServerSelector>> filteredServersLookup =
         toolChest.filterSegments(query, serversLookup);
 
+    // Absent Optional indicate a null rangeset which represents range cannot be determined
+    final Map<String, Optional<RangeSet<String>>> dimensionRangeMap = Maps.newHashMap();
+    final DimFilter filter = query.getFilter();
+
+    // Filter unneeded chunks based on partition dimension
     for (TimelineObjectHolder<String, ServerSelector> holder : filteredServersLookup) {
       for (PartitionChunk<ServerSelector> chunk : holder.getObject()) {
         ServerSelector selector = chunk.getObject();
-        final SegmentDescriptor descriptor = new SegmentDescriptor(
-            holder.getInterval(), holder.getVersion(), chunk.getChunkNumber()
-        );
+        boolean include = true;
+
+        if (filter != null) {
+          Map<String, Range<String>> domain = selector.getSegment().getShardSpec().getDomain();
+          for (Map.Entry<String, Range<String>> entry : domain.entrySet()) {
+            Optional<RangeSet<String>> optFilterRangeSet = dimensionRangeMap.get(entry.getKey());
+            if (optFilterRangeSet == null) {
+              RangeSet<String> filterRangeSet = filter.getDimensionRangeSet(entry.getKey());
+              optFilterRangeSet = filterRangeSet == null ? Optional.<RangeSet<String>>absent() : Optional.of(filterRangeSet);
+              dimensionRangeMap.put(entry.getKey(), optFilterRangeSet);
+            }
+            if (optFilterRangeSet.isPresent() && optFilterRangeSet.get().subRangeSet(entry.getValue()).isEmpty()) {
```

Comment:
oops, nevermind, I was thinking of it's always non-null. It might not always be present. nevermind :)